ToolBox/BovespaDownloader/BovespaDataDownloader.cs

```diff
@@ -0,0 +1,456 @@
+                    zip.Dispose();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Aggregates a list of ticks at the requested resolution
+        /// </summary>
+        /// <param name="symbol">Symbol</param>
+        /// <param name="ticks">Input tick data</param>
+        /// <param name="resolution">Output resolution</param>
+        /// <returns>Enumerable of trade bars or quote bars</returns>
+        private static IEnumerable<BaseData> AggregateTicks(Symbol symbol, IEnumerable<BaseData> data, Resolution resolution)
+        {
+            var ticks = data.Cast<Tick>().OrderBy(t => t.Time).ToList();
+
+            if (ticks.All(t => t.TickType == TickType.Trade))
+            {
+                return
+                    (from t in ticks
+                     group t by t.Time.RoundDown(resolution.ToTimeSpan())
+                         into g
+                         select new TradeBar
+                         {
+                             DataType = MarketDataType.TradeBar,
+                             Symbol = symbol,
+                             Time = g.Key,
+                             Open = g.First().LastPrice,
+                             High = g.Max(t => t.LastPrice),
+                             Low = g.Min(t => t.LastPrice),
+                             Close = g.Last().LastPrice,
+                             Volume = g.Sum(t => t.Quantity),
+                         });
+            }
+            else
+            {
+                return 
+                    (from t in ticks
+                     group t by t.Time.RoundDown(resolution.ToTimeSpan()) into g
+                     
+                     select new QuoteBar
+                     {
+                         DataType = MarketDataType.QuoteBar,
+                         Symbol = symbol,
+                         //Time = g.Key,
+                         //AvgBidSize = (long)g.Where(t => t.BidSize > 0).Average(t => t.BidSize),
+                         //AvgAskSize = (long)g.Where(t => t.AskSize > 0).Average(t => t.AskSize),
+                         //Bid = new Bar
+                         //(
+                         //    g.Where(t => t.BidSize > 0).First().LastPrice,
+                         //    g.Where(t => t.BidSize > 0).Max(t => t.LastPrice),
+                         //    g.Where(t => t.BidSize > 0).Min(t => t.LastPrice),
+                         //    g.Where(t => t.BidSize > 0).Last().LastPrice
+                         //),
+                         //Ask = new Bar
+                         //(
+                         //    g.Where(t => t.AskSize > 0).First().LastPrice,
+                         //    g.Where(t => t.AskSize > 0).Max(t => t.LastPrice),
+                         //    g.Where(t => t.AskSize > 0).Min(t => t.LastPrice),
+                         //    g.Where(t => t.AskSize > 0).Last().LastPrice
+                         //),         
+                     });
```

Comment:
If by optimal we're talking about performance, then not entirely. LINQ is a great language feature but does have some overhead given all the delegates/method calls. The `QuoteBar` class has an `Update` [method](https://github.com/QuantConnect/Lean/blob/master/Common/Data/Market/QuoteBar.cs#L223) that could be called in a loop. This would probably be the most efficient, straightforward way.
```C#
var resolutionIncrement = resolution.ToTimeSpan();
var grouped = ticks.GroupBy(t => t.Time.RoundDown(resolutionIncrement);
foreach (var g in grouped)
{
    // maybe we can make another constructor that just accepts time,symbol and resolution and
    // calls this other constructor with these values. this would make the aggregation case easier
    var quoteBar = new QuoteBar(t.Time, t.Symbol, null, 0, null, 0, resolutionIncrement);
    foreach (var tick in g)
    {
        quoteBar.Update(0, tick.BidPrice, tick.AskPrice, 0, tick.BidSize, tick.AskSize);
    }
    yield return quoteBar;
}
```

I just wrote this up in this window, so it may not compile :/ apologies, but I think it expresses the intent. The performance issue with using LINQ to do this aggregation is that we end up looping about 6x more then we need to, as each one of the `Max()`, `Min()`, and `Last()` calls will loop over the entire grouping.

I try to use the `Update` methods when possible since we can write a test against it, and if everyone uses it then we know everyone is aggregating with the same rules.